File: 02-linked-lists/20250727-beginner/ts/solution.ts
// Node class represents a single node in the linked list
class Node {
    data: number;
    next: Node | null;

    constructor(data: number) {
        this.data = data;
        this.next = null;
    }
}

// LinkedList class
export class LinkedList {
    private head: Node | null;

    constructor() {
        this.head = null;
    }

    // TODO: Implement the append method
    // append adds a new node with the given data at the end of the list
    append(data: number): void {
        // TODO: Implement this method
        const newNode = new Node(data)
        if (this.head === null) {
            this.head = newNode
            return
        }
        let current = this.head
        while (current.next !== null) {
            current = current.next
        }
        current.next = newNode
    }

    // TODO: Implement the findMiddle method
    // findMiddle returns the data of the middle node in the linked list
    // For lists with even number of elements, return the first of the two middle elements
    findMiddle(): { data: number; found: boolean } {
        // TODO: Implement this method
        // Return the middle element's data and found=true if list is not empty
        // Return data=0 and found=false if list is empty
        if (this.head == null || this.size() === 0) {
            return { data: 0, found: false };
        }
        const size = this.size()
        const target = size % 2 === 0 ? Math.floor(size / 2) - 1 : Math.floor(size / 2)
        let current = this.head
        for (let i = 0; i < target; i++) {
            current = current.next as Node
        }
        return { data: current.data, found: true };
    }

    // TODO: Implement the toArray method
    // toArray returns all elements in the list as an array
    toArray(): number[] {
        // TODO: Implement this method
        let out = []
        let current = this.head
        while (current != null) {
            out.push(current.data)
            current = current.next
        }

        return out;
    }

    // TODO: Implement the size method
    // size returns the total number of nodes in the list
    size(): number {
        // TODO: Implement this method
        let count = 0
        let current = this.head
        while (current != null) {
            count++
            current = current.next
        }
        return count;
    }
}

File: 02-linked-lists/20250727-beginner/go/solution.go
package main

// Node represents a single node in the linked list
type Node struct {
	data int
	next *Node
}

// LinkedList represents the linked list structure
type LinkedList struct {
	head *Node
}

// TODO: Implement the append method
// append adds a new node with the given data at the end of the list
func (ll *LinkedList) append(data int) {
	// TODO: Implement this method
	newNode := &Node{data: data, next: nil}
	if ll.head == nil {
		ll.head = newNode
		return
	}
	current := ll.head
	for current.next != nil {
		current = current.next
	}
	current.next = newNode
}

// TODO: Implement the findMiddle method
// findMiddle returns the data of the middle node in the linked list
// For lists with even number of elements, return the first of the two middle elements
func (ll *LinkedList) findMiddle() (int, bool) {
	// TODO: Implement this method
	// Return the middle element's data and true if list is not empty
	// Return 0 and false if list is empty
	if ll.head == nil {
		return 0, false
	}
	size := ll.size()
	if size == 1 {
		return ll.head.data, true
	}
	target := (size - 1) / 2
	current := ll.head
	for range target {
		current = current.next
	}
	return current.data, true

}

// TODO: Implement the toSlice method
// toSlice returns all elements in the list as a slice
func (ll *LinkedList) toSlice() []int {
	// TODO: Implement this method
	if ll.head == nil {
		return []int{}
	}
	out := make([]int, 0)
	current := ll.head
	for current != nil {
		out = append(out, current.data)
		current = current.next
	}
	return out
}

// TODO: Implement the size method
// size returns the total number of nodes in the list
func (ll *LinkedList) size() int {
	// TODO: Implement this method
	count := 0
	current := ll.head
	for current != nil {
		count++
		current = current.next
	}
	return count
}

